### ISSUE TITLE
Search

### ISSUE THREAD

**Issue by gerbrent** [JupiterBroadcasting/jupiterbroadcasting.com#26]
figure out a good way to integrate search. Clientside like Lunr.js will probably not perform due to index size.

@theZMC recommends: 
> Since golang seems to be a theme here (which it should be), maybe [zinc](https://github.com/zinclabs/zinc) would fit the bill?

**Comment by elreydetoda** [comment]
If zinc is used (from my understanding) it looks like it'll need to be implemented on the server side (at least for the server which holds all the references to the objects/records).

Also, of note it appears that zinc is still in beta ([src](https://docs.zincsearch.com/#project-status)):

> Project Status:
ZincSearch is in Pre GA (General Availability) and will be marked as production ready at v1.0.0 .

So, it's possible they'll have major breaking changes before the 1.0 release, which means we'll need to make sure we pin the version and read the changelog before upgrading to see if anything's going to break.

While it's not a self-hosted option (and will probably cost money because of how many episodes JB has), as a temporary solution, we could use [algolia](https://www.algolia.com). I've used them before, and overall it was pretty easy. I've actually got a GH action for doing CI with hugo content as well: https://github.com/Climate-Refugee-Stories/crs-website/blob/c82f394a620b4631bb43de5ca4433d33a51bb292/.github/workflows/cd.yml#L91-L126
(figure we probably won't go with this, but just figured I'd mention it).

**Comment by elreydetoda** [comment]
Meta: BTW, @gerbrent  you might want to add a "JB - action needed" tag to this issue since it's discussing cost of running an extra service on a server specifically for search, and if that's something they want to contemplate (because that'll be another service to maintain).

**Comment by reesericci** [comment]
Typesense might be a good option

**Comment by ironicbadger** [comment]
@realorangeone usually has some pretty strong opinions on search.

**Comment by gerbrent** [comment]
I have opinions aswell, from a functionality and end-user perspective. 

The search results at [notes.jupiterbroadcasting.com](https://notes.jupiterbroadcasting.com/) is not at all to my liking pretty much every single time I try to use it, which is generally answering a question like "I remember we mentioned that in the last few months, lets see which episode that was from" - This generally gives results sorted by "relevance" which never gets me what I want (and yet I keep trying.....)

I would far prefer chronologically sorted search results.

- see https://github.com/selfhostedshow/show-notes/issues/16

Also on a slow connection, the UX of that current search - the present-results-as-pop-up-in-search-bar behaviour isn't obvious for quite some time till the results load. An annoying UX experience, and slow enough to make me wonder more than once "is this working?"

**Comment by theZMC** [comment]
I'm willing to start putting some serious development work into this. Some clarifying questions:
- Do we need full text search?
- Any plans for some sort of transcription process (automatic or manual) and if so is there a timeline for when that would be in place?

**Comment by ironicbadger** [comment]
See the search at `notes.jupiterbroadcasting.com` - that's the type of thing *I* think is needed. @ChrisLAS @noblepayne or @gerbrent feel free to jump in here.

**Comment by gerbrent** [comment]
see here why search via recency is not supported nor desired by mkdocs:

https://github.com/selfhostedshow/show-notes/issues/16#issuecomment-1216728294

**Comment by RealOrangeOne** [comment]
I agree lunr probably isn't ideal, as the index will be huge. [I've written client-side search with Hugo](https://theorangeone.net/posts/hugo-website-search/), and it's very simple, but the index may be large given the show history

mkdocs's search is lunr-based. The issue with mkdocs is that the pages have no sense of date, as opposed to being a technological issue. Hugo however _does_ have dates as a concept, so could be done.

For search, I suspect we'd want want something server-side to do it. For ease (of local dev and hosting), scraping the content into sqlite and using its fulltext search would probably be very simple, very powerful and scalable.

Elasticsearch etc are definitely options, but they're very heavy for what we need. As are hosted tools like Algolia, but given the name of one of our shows, that's a less desirable option.

**Comment by ironicbadger** [comment]
Could we run some mock ups with elastic and get a sense of just how heavy? We have the infra to do it I'd wager.

**Comment by RealOrangeOne** [comment]
It's not just heavy in terms of resource. It also makes local development much more of a pain, not to mention is more complex to setup and work with anyway. The container alone is ~550mb compressed.

**Comment by theZMC** [comment]
> It's not just heavy in terms of resource. It also makes local development much more of a pain, not to mention is more complex to setup and work with anyway. The container alone is ~550mb compressed.

Unfortunately when it comes to search, I think it's the classic pick two between fast, good, and inexpensive. Though I do agree that full fat elastic is a bit too heavy-handed for our needs.

**Comment by CGBassPlayer** [comment]
So I just found a tool that might be worth using if search is still something we are after. Its called [Pagefind](https://pagefind.app/) and it is a single binary that indexes the site after it is built. There is a video on the home page of their site showing how it works and a basic example.

Its also written in :tada: ***Rust*** :tada:)

**Comment by elreydetoda** [comment]
That looks pretty awesome! It's nice that we could just bundle that in an artifact as well. It just comes with the new site build! 🥳

**Comment by gerbrent** [comment]
My big question - can results be sorted by date/recency? I see Pagefind has the concept of "date"

![image](https://user-images.githubusercontent.com/7018928/189196301-b0cc90ed-c208-4933-a690-524bc03e5396.png)

**Comment by gerbrent** [comment]
> Pagefind can be configured to search across multiple sites, merging results and filters into a single response. Multisite search configuration happens entirely in the browser, by pointing one Pagefind instance at multiple search bundles.
>
> The following examples reflect Pagefind running on a website at blog.example.com that wants to include pages from docs.example.com in the search results.

https://pagefind.app/docs/multisite/

That can be very handy for the JB Archive (a distinct hugo instance)

**Comment by FlakM** [comment]
Hello all! Have you considered https://www.meilisearch.com/ it's also an open source project with a valid source  of income (they have recently received 15M round o founding). It is very easy to deploy and used LMDB and rust underneath. I'd be more than happy to write backend RSS watcher and some mockups for front end.
As for costs times are crazy but I think I can commit to covering a year of runtime and on call support as value for value 🥰

**Comment by gerbrent** [comment]
oh wow, very generous @FlakM !!!

I'll be curious to hear what others think of MeiliSearch - def worth considering!

**Comment by FlakM** [comment]
I've been recently reviewing alternatives for more traditional ELK stack and hosted options for my employer. Meilisearch has come up on this week in rust so I have also looked into it. Here are some reasons why I think it would be a good fit here:

- It uses a very solid backing technologies - ie [LMDB](https://docs.meilisearch.com/learn/advanced/storage.html#lmdb) which has been designed as a embedded database for openldap by very smart people. If you prefer podcasts here is a great [episode about it](https://www.se-radio.net/2019/06/episode-371-howard-chu-on-the-lightning-memory-mapped-database-lmdb/).
- It has a rest API so it could be used without any other backend services apart from the component that will keep data in sync (and maybe some nginx to add some rate limiting/tls etc)
- It is dead simple to deploy and maintain - just a single container
- It has a front-end code already written so including it is also very simple
- It has all of those nice features like typo safety, synonyms etc
- It is blazingly fast :rocket: :crab: 

For your convenience, I've deployed a sample service and loaded the index with contents of all feeds RSS its available [here](http://139.144.189.54:7700/) (BTW it is a proper use of  Linode credits) secret key is `MASTER_KEY`. Keep in mind that it is a result of a fast and dirty effort. For a full-blown index, I think it would be useful to also add transcription (I've experience with deep speech so probably not a big problem) and more complete show notes. The current showcase version of code loading data from the RSS feed is available [here](https://github.com/FlakM/jupiter-search)

**Comment by elreydetoda** [comment]
@kylepotts suggests [start of convo](https://matrix.to/#/!rDlEtonuSythLzmxQL:jupiterbroadcasting.com/$VGnfifOm2JYLKZAwMxM-vOA-fi9yhLXkUqTYcwKeCNc) & [end of convo](https://matrix.to/#/!rDlEtonuSythLzmxQL:jupiterbroadcasting.com/$Sk3S84tGKcy6wAPwUb69d-skSIv2LuK2qsXKn5L7Nkg):
> What things have we tried for searching transcriptions? I wonder if taking the output of the transcription and putting it inside something like ElasticSearch/Opensearch and exposing it via an API is overkill? Or if a product like that already exists. Definitely will require a unique way to have a "dynamic" results page in Hugo from where you search.
